backend/rust/models/src/llama/llama.rs

```diff
@@ -0,0 +1,720 @@
+
+    // the last two dims need to be (n_ctx, n_hstate)
+    let q = rotary_encoder.forward(q.swap_dims(1, 2)) * scale;
+    let k = rotary_encoder.forward(k.swap_dims(1, 2)) * scale;
+    let v = v.swap_dims(1, 2);
+
+    // compute value weightings
+    let qk = q.matmul(k.transpose());
+
+    // apply mask
+    let qk = if let Some(mask) = mask {
+        qk + mask.slice([0..n_qctx, 0..n_ctx]).unsqueeze::<4>()
+    } else {
+        qk
+    };
+
+    // normalize value weightings
+    let w = softmax(qk, 3);
+    let o = w.matmul(v).swap_dims(1, 2).flatten(2, 3);
+
+    return o;
+}
+
+/// For a tensor of size (n_batch, n_ctx, n_kv_head, n_hstate), repeats the head keys or values in an interleaving manner so that the number
+/// of heads is effectively multiplied by n_repeat
+fn repeat_kv<B: Backend>(x: Tensor<B, 4>, n_repeat: usize) -> Tensor<B, 4> {
+    if n_repeat > 1 {
+        let [n_batch, n_ctx, n_kv_head, n_hstate] = x.dims();
+        x.repeat(3, n_repeat)
+            .reshape([n_batch, n_ctx, n_kv_head * n_repeat, n_hstate])
+    } else {
+        x
+    }
+}
+
+/// Generates a strictly upper triangular matrix filled with -inf that when added to an attention weight matrix prevents
+/// vectors from attending to other vectors further in the sequence, preventing future information from flowing into the past
+pub fn attn_decoder_mask<B: Backend>(seq_length: usize) -> Tensor<B, 2> {
+    let mut mask = Tensor::<B, 2>::zeros([seq_length, seq_length]);
+
+    for i in 0..(seq_length - 1) {
+        let values = Tensor::<B, 2>::zeros([1, seq_length - (i + 1)]).add_scalar(NEG_INFINITY);
+        mask = mask.slice_assign([i..i + 1, i + 1..seq_length], values);
+    }
+
+    return mask;
+}
+
+#[derive(Config, Debug)]
+pub struct RotaryEncodingConfig {
+    max_sequence_length: usize,
+    state_size: usize,
+    theta: f64,
+}
+
+impl RotaryEncodingConfig {
+    pub fn init<B: Backend>(&self) -> RotaryEncoding<B> {
+        assert!(self.state_size % 2 == 0, "Head dims must be even.");
+        assert!(self.theta > 0.0, "Theta must be positive.");
+
+        let half_state_size = self.state_size / 2;
+
+        let arange_m = Tensor::from_floats([[1.0, 0.0, 0.0, 1.0], [0.0, -1.0, 1.0, 0.0]]).into();
+
+        let inv_freq = powto(
+            self.theta,
+            Tensor::arange(0..half_state_size).float() * (2.0 / self.state_size as f64),
+        )
+        .powf(-1.0);
+
+        let periods = Tensor::arange(0..self.max_sequence_length)
+            .float()
+            .unsqueeze::<2>()
+            .transpose()
+            .repeat(1, half_state_size)
+            * inv_freq.unsqueeze();
+
+        let p_cos = periods.clone().cos();
+        let p_sin = periods.sin();
+        let tensor=Tensor::cat(vec![p_cos, p_sin], 1);
+        
+        let tensor2=tensor.reshape([self.max_sequence_length,2,half_state_size]);
+
+        let tensor3=tensor2.transpose();
+
+        let tensor41=tensor3.repeat(2, 2);
```

Comment:
@lu-zero Here, I am going to use `wgpu` backend instead of `tch`. However, I the repeat function here only support 2 dimensions tensor, (Can only repeat dimension with dim=1) https://github.com/Tracel-AI/burn/blob/b86bc5876149bd73bc59cb5197fd3ee8b92509d4/burn-tensor/src/tensor/ops/tensor.rs#L222C7-L222C7. 

I have been tried several solutions, like use `swap_dims` and `flattern` these internal function of Tensor, but here hard to say it is correct and also causes other issues. Is there a better example for this?

Reply:
asking upstream probably it is the best route (sorry for the belated reply, I got very busy and the message got lost in the mailbox)

Reply:
No worries, thanks for your support. I will continue to work on this one after  I applied PhD successfully. Currently, sooo busy. But I still want to get this PR to merged.

Reply:
Once you are more free please contact me, probably a good deal of the issues will be ironed out by upstream meanwhile :)